L1Trigger/TrackFindingTMTT/src/L1KalmanComb.cc

```diff
@@ -0,0 +1,1828 @@
+
+        std::vector<const Stub *> stubs_for_cls;
+        stubs_for_cls.push_back(layer_stubs.at(i));
+
+#ifdef MERGE_STUBS
+        while (layer_stubs.at(i) != layer_stubs.back()) {
+          if (isOverlap(layer_stubs.at(i), layer_stubs.at(i + 1), TYPE_NORMAL)) {
+            stubs_for_cls.push_back(layer_stubs.at(i + 1));
+            if (getSettings()->kalmanFillInternalHists())
+              hnmergeStub_->Fill(0);
+            i++;
+          } else
+            break;
+        }
+#endif
+
+        if (getSettings()->kalmanFillInternalHists()) {
+          if (tpa && tpa->useForAlgEff()) {
+            if (stubs_for_cls.size() > 1) {
+              std::set<const TP *> s_tps = stubs_for_cls.at(0)->assocTPs();
+              if (s_tps.find(tpa) != s_tps.end()) {
+                const Stub *sa = stubs_for_cls.front();
+                const Stub *sb = stubs_for_cls.back();
+
+                double drphi = fabs(sa->r() * reco::deltaPhi(sa->phi(), sectorPhi()) -
+                                    sb->r() * reco::deltaPhi(sb->phi(), sectorPhi()));
+                double dz = fabs(sa->z() - sb->z());
+                double dr = fabs(sa->r() - sb->r());
+                TString hname;
+                if (LayerId[j_layer] < 10) {
+                  hname = Form("hBarrelStubMaxDistanceLayer%02d", LayerId[j_layer]);
+
+                  if (hBarrelStubMaxDistanceMap.find(hname) == hBarrelStubMaxDistanceMap.end()) {
+                    cout << hname << " does not exist." << endl;
+                  } else {
+                    hBarrelStubMaxDistanceMap[hname]->Fill(drphi, dz);
+                  }
+                } else {
+                  hname = Form("hEndcapStubMaxDistanceRing%02d", sa->endcapRing());
+
+                  if (hEndcapStubMaxDistanceMap.find(hname) == hEndcapStubMaxDistanceMap.end()) {
+                    cout << hname << " does not exist." << endl;
+                  } else {
+                    hEndcapStubMaxDistanceMap[hname]->Fill(drphi, dr);
+                  }
+                }
+              }
+            }
+          }
+        }
+
+        // dl error now disabled
+        StubCluster *stbcl = new StubCluster(stubs_for_cls, sectorPhi(), 0);
+        stbcl_list_.push_back(stbcl);
+        stubcls.push_back(stbcl);
+
+        if (getSettings()->kalmanFillInternalHists()) {
+          if (!stbcl->barrel()) {
+            TString hname = Form("hphiErrorRatioRing%d", stbcl->endcapRing());
+            if (hphiErrorRatioMap.find(hname) == hphiErrorRatioMap.end()) {
+              cout << hname << " does not exist." << endl;
+            } else {
+              hphiErrorRatioMap[hname]->Fill(fabs(stbcl->deltai() + 0.5), fabs(stbcl->dphi_dr()) / stbcl->dphi_dl());
+            }
+          }
+        }
+      }
+    }
+    if (getSettings()->kalmanFillInternalHists()) {
+      if (tpa && tpa->useForAlgEff()) {
+        hTrackEta_->Fill(tpa->eta());
+        static set<const TP *> set_tp;
```

Comment:
Please avoid function static, it is thread hostile.

Reply:
The internal histograms have all been deleted from the Kalman Fit code.